### PR TITLE
refactor(qos): Operational changes

### PR DESCRIPTION
**Pull request by robzienert**
A couple handfuls of refactors & fixes from post-merge conversations with various people. They're all pretty small, probably best reviewed commit-by-commit.

### CODE REVIEW

File: orca-qos/src/main/kotlin/com/netflix/spinnaker/orca/qos/ExecutionBufferActuator.kt [hunk 1]
```diff
@@ -25,21 +26,27 @@ import com.netflix.spinnaker.orca.qos.BufferState.ACTIVE
...
 import org.springframework.context.event.EventListener
+import org.springframework.core.env.Environment
 import org.springframework.stereotype.Component
 
 /**
  * Determines if an execution should be buffered.
  */
 @Component
 class ExecutionBufferActuator(
+  private val environment: Environment,
```
Comment (robzienert) [inline orca-qos/src/main/kotlin/com/netflix/spinnaker/orca/qos/ExecutionBufferActuator.kt]: I went with `Environment` so we can drive learning mode from FP.

File: orca-qos/src/main/kotlin/com/netflix/spinnaker/orca/qos/ExecutionBufferActuator.kt [hunk 2]
```diff
@@ -25,21 +26,27 @@ import com.netflix.spinnaker.orca.qos.BufferState.ACTIVE
...
+  private val registry: Registry,
   policies: List<BufferPolicy>
 ) {
 
   private val log = LoggerFactory.getLogger(ExecutionBufferActuator::class.java)
 
   private val orderedPolicies = policies.sortedByDescending { it.order }.toList()
 
+  private val bufferedId = registry.createId("qos.executionsBuffered")
+  private val elapsedTimeId = registry.createId("qos.promoter.elapsedTimeMs")
```
Comment (ajordens) [inline orca-qos/src/main/kotlin/com/netflix/spinnaker/orca/qos/ExecutionBufferActuator.kt]: May not want to include the `Ms` in the metric id

File: orca-qos/src/main/kotlin/com/netflix/spinnaker/orca/qos/ExecutionPromoter.kt [hunk 3]
```diff
@@ -33,48 +38,70 @@ interface ExecutionPromoter
...
-  private val policies: List<PromotionPolicy>
-) : ExecutionPromoter {
+  private val policies: List<PromotionPolicy>,
+  private val registry: Registry
+) : ExecutionPromoter, ApplicationListener<RemoteStatusChangedEvent> {
 
   private val log = LoggerFactory.getLogger(ExecutionPromoter::class.java)
+  private val discoveryActivated = AtomicBoolean()
+
+  private val elapsedTimeId = registry.createId("qos.promoter.elapsedTimeMs")
```
Comment (ajordens) [inline orca-qos/src/main/kotlin/com/netflix/spinnaker/orca/qos/ExecutionPromoter.kt]: same comment here about `Ms` (just not normally what we do)

File: orca-qos/src/main/kotlin/com/netflix/spinnaker/orca/qos/ExecutionPromoter.kt [hunk 4]
```diff
@@ -33,48 +38,70 @@ interface ExecutionPromoter
...
-        // result.finalized==true or the end of the list, then be able to pass all contributing source & reason pairs
-        // into a log, with a zipped summary that would be saved into an execution's system notifications.
-        var lastResult: PromotionResult? = null
-        var candidates = it
-        policies.forEach { policy ->
-          val result = policy.apply(candidates)
-          if (result.finalized) {
-            return@let result
+    if (!discoveryActivated.get()) {
+      return
```
Comment (ajordens) [inline orca-qos/src/main/kotlin/com/netflix/spinnaker/orca/qos/ExecutionPromoter.kt]: 👍

File: orca-qos/src/main/kotlin/com/netflix/spinnaker/orca/qos/bufferpolicy/EnqueueDeckOrchestrationsBufferPolicy.kt [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+ * Deck-initiated orchestrations should always be enqueued. This buffer policy cannot be disabled.
+ */
+@Component
+class EnqueueDeckOrchestrationsBufferPolicy : BufferPolicy {
+  override fun apply(execution: Execution): BufferResult {
+    if (execution.type == ORCHESTRATION && execution.origin == "deck") {
+      return BufferResult(
+        action = ENQUEUE,
+        force = true,
+        reason = "Deck-initiated orchestrations are always enqueued"
```
Comment (ajordens) [inline orca-qos/src/main/kotlin/com/netflix/spinnaker/orca/qos/bufferpolicy/EnqueueDeckOrchestrationsBufferPolicy.kt]: 🌷

**Comment by ajordens** [comment]
Thanks for making the tweaks @robzienert

**Pull request merged**